### PR TITLE
Cleanup: agreementMapper, provinces

### PR DESCRIPTION
**Pull request by Zemurin**

### CODE REVIEW

File: EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h [hunk 1]
```diff
@@ -1,49 +1,49 @@
...
+	explicit ProvinceHistory(std::istream& theStream);
+
+	void updatePopRatioCulture(const std::string& oldCultureName, const std::string& neoCultureName, const std::string& superRegion);
+	void buildPopRatios(double assimilationFactor);
+	void setStartingCulture(const std::string& culture) { startingCulture = culture; }
+	void setStartingReligion(const std::string& religion) { startingCulture = religion; }
+
+	[[nodiscard]] std::optional<date> getFirstOwnedDate() const;
+	[[nodiscard]] bool hasOriginalCulture() const;
+	[[nodiscard]] const auto& getOriginalCulture() const { return startingCulture; }
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:24]: Add tests

File: EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h [hunk 2]
```diff
@@ -1,49 +1,49 @@
...
+	[[nodiscard]] bool wasColonized() const;
+	[[nodiscard]] bool hasInitializedHistory() const { return !religionHistory.empty() && !cultureHistory.empty(); }
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:26]: Ditto

File: EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h [hunk 3]
```diff
@@ -1,49 +1,49 @@
...
-		double originalManpower = 0;
-	};
-}
+class ProvinceHistory: commonItems::parser
+{
+  public:
+	ProvinceHistory() = default;
+	explicit ProvinceHistory(std::istream& theStream);
+
+	void updatePopRatioCulture(const std::string& oldCultureName, const std::string& neoCultureName, const std::string& superRegion);
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:17]: Add tests

File: EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h [hunk 4]
```diff
@@ -1,49 +1,49 @@
...
+	void buildPopRatios(double assimilationFactor);
+	void setStartingCulture(const std::string& culture) { startingCulture = culture; }
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:19]: Add tests

File: EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h [hunk 5]
```diff
@@ -1,49 +1,49 @@
...
+	void setStartingReligion(const std::string& religion) { startingCulture = religion; }
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:20]: Add tests

Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/ProvinceHistory.h:20]: Also startingCulture -> startingReligion

File: EU4toV2/Source/EU4World/Provinces/EU4Province.h [hunk 6]
```diff
@@ -1,17 +1,16 @@
...
 #include "ProvinceStats.h"
-#include "../../Mappers/Buildings/Buildings.h"
-#include "../Modifiers/Modifiers.h"
+#include "Buildings/Buildings.h"
+#include "Modifiers/Modifiers.h"
 #include "Parser.h"
-#include "../../Mappers/SuperGroupMapper/SuperGroupMapper.h"
-#include "../Regions/Regions.h"
+#include "SuperGroupMapper/SuperGroupMapper.h"
+#include "Regions/Regions.h"
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/EU4Province.h:13]: Many of the pre-existing functions are missing tests. Even for the simple ones, tests should be added as a demonstration of use.

File: EU4toV2/Source/EU4World/Provinces/EU4Province.cpp [hunk 7]
```diff
@@ -136,44 +25,119 @@ EU4::Province::Province(const std::string& numString, std::istream& theStream)
...
-	if (buildings.hasBuilding(building)) return true;
-	return false;
+	return buildings.hasBuilding(building);
 }
 
 bool EU4::Province::hasGreatProject(const std::string& greatProject) const
 {
-	if (greatProjects.count(greatProject) > 0) return true;
-	return false;
+	return greatProjects.count(greatProject);
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/EU4Province.cpp:131]: As elsewhere, count -> contains if possible

File: EU4toV2/Source/EU4World/Provinces/Provinces.h [hunk 8]
```diff
@@ -1,36 +1,36 @@
...
-	{
-	public:
-		explicit Provinces(std::istream& theStream);
-		
-		[[nodiscard]] auto geTotalProvinceWeights() const { return totalProvinceWeights; }
-		[[nodiscard]] const auto& getAllProvinces() const { return provinces; }
-		std::shared_ptr<Province> getProvince(int provinceNumber);
-		
-		void determineTotalProvinceWeights(const Configuration& configuration);
+	[[nodiscard]] auto geTotalProvinceWeights() const { return totalProvinceWeights; }
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/Provinces.h:21]: geTotalProvinceWeights -> getTotalProvinceWeights

Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Provinces/Provinces.h:21]: Add tests

File: EU4toV2/Source/Mappers/AgreementMapper/AgreementMapper.h [hunk 9]
```diff
@@ -1,32 +1,31 @@
...
-		[[nodiscard]] auto isAgreementInDoublesiders(const std::string& subject) const { return doublesiders.count(subject) > 0; }
-		[[nodiscard]] auto isAgreementInTributaries(const std::string& subject) const { return tributaries.count(subject) > 0; }
-		[[nodiscard]] auto isAgreementInVassals(const std::string& subject) const { return vassals.count(subject) > 0; }
+class AgreementMapper: commonItems::parser
+{
+  public:
+	AgreementMapper();
+	explicit AgreementMapper(std::istream& theStream);
+
+	[[nodiscard]] auto isAgreementInColonies(const std::string& subject) const { return colonies.count(subject); }
```
Comment (Idhrendur) [inline EU4toV2/Source/Mappers/AgreementMapper/AgreementMapper.h:14]: As elsewhere, count -> contains if possible. Same for following functions.

**Review by Idhrendur — APPROVED** [review]

File: EU4toV2/Source/EU4World/Provinces/ProvinceBuildings.h [hunk 10]
```diff
@@ -1,23 +1,24 @@
...
-		
-		[[nodiscard]] auto hasBuilding(const std::string& building) const { return buildings.count(building) > 0; }
-		[[nodiscard]] const auto& getBuildings() const { return buildings; }
+class ProvinceBuildings: commonItems::parser
+{
+  public:
+	ProvinceBuildings() = default;
+	explicit ProvinceBuildings(std::istream& theStream);
+
+	[[nodiscard]] auto hasBuilding(const std::string& building) const { return buildings.count(building) > 0; }
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Provinces/ProvinceBuildings.h:14]: No, not yet for either this or cks. That's a PR of its own when time is right.

File: EU4toV2/Source/EU4World/Provinces/Provinces.cpp [hunk 11]
```diff
@@ -1,35 +1,40 @@
...
 	});
-	registerRegex("[a-zA-Z0-9_\\.:]+", commonItems::ignoreItem);
-
-	parseStream(theStream);
-	clearRegisteredKeywords();
+	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
 }
 
-std::shared_ptr<EU4::Province> EU4::Provinces::getProvince(const int provinceNumber)
+const std::shared_ptr<EU4::Province>& EU4::Provinces::getProvince(const int provinceNumber) const
```
Comment (IhateTrains) [inline EU4toV2/Source/EU4World/Provinces/Provinces.cpp:22]: Can't the return type be auto?

Comment (Zemurin) [reply EU4toV2/Source/EU4World/Provinces/Provinces.cpp:22]: image
image
You can, but it's fuzzy about having not being defined although it obviously is, and I don't see immediately what's the return type when looking at the cpp file.

File: EU4toV2/Source/EU4World/Provinces/Provinces.h (hunk 8 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Provinces/Provinces.h:21]: Always wondered about that one.
https://github.com/ParadoxGameConverters/EU4toVic2/commit/a22ae045aba36a33266870549121a3fce6533f61
Figured you wanted it to stand out or something. Would be a shame to change it, now that it's become bedrock of converter history.

Comment (Zemurin) [reply EU4toV2/Source/EU4World/Provinces/Provinces.h:21]: Seriously now:
image
I don't even comprehend this code. To test it I'd have to understand it.

I'll see if I can break it down into something human-readible.

File: EU4toV2/Source/EU4World/Provinces/Provinces.cpp (hunk 11 above)
Comment (IhateTrains) [reply EU4toV2/Source/EU4World/Provinces/Provinces.cpp:22]: On the screenshot, did you also change it to auto in the definition?

Comment (IhateTrains) [reply EU4toV2/Source/EU4World/Provinces/Provinces.cpp:22]: That's fine, auto is mostly for our convenience anyway.

**Review by IhateTrains — APPROVED** [review]

File: EU4toV2/Source/EU4World/Provinces/Provinces.h (hunk 8 above)
Comment (Idhrendur) [reply EU4toV2/Source/EU4World/Provinces/Provinces.h]: Ha! Should have known it was my own fault. It usually is.

Comment (Idhrendur) [reply EU4toV2/Source/EU4World/Provinces/Provinces.h]: Oh yes, that. It's one of those bits I meant to just keep refactoring bit by bit. Skip for now, but maybe open an issue to come back to it.

I suspect some bits are trying to replicate EU4 logic for things we can read. Maybe with some modifiers. Any such modifiers should be moved from code to a configurable.

**Pull request merged**